### PR TITLE
[Resource] Add Host

### PR DESCRIPTION
**Pull request by michaelkad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMIHostBasic
--- PASS: TestAccIBMIHostBasic (83.42s)
PASS
...
```

### CODE REVIEW

File: ibm/flex/structures_test.go [hunk 1]
```diff
@@ -81,6 +82,10 @@ func TestStringifyList(t *testing.T) {
...
 
 func TestStringifyMap(t *testing.T) {
-	var foo interface{} = map[string]interface{}{"foo": "bar"}
-	assert.Equal(t, `{"foo":"bar"}`, Stringify(foo))
+	var foo interface{} = map[string]interface{}{"foo": "bar", "cl": "car"}
+	x := Stringify(foo)
+	fmt.Println(x)
+	fmt.Println(x[1], x[4], "here")
+
+	// assert.Equal(t, `{"cl":"car", "foo":"bar"}`, Stringify(foo))
```
Comment (ismirlia) [inline ibm/flex/structures_test.go:90]: what is this change?

File: ibm/service/power/resource_ibm_pi_host.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+func resourceIBMPIHostUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(conns.ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	cloudInstanceID, hostID, err := splitID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	displayName := d.Get(Arg_Host + ".0").(map[string]interface{})[Attr_DisplayName].(string)
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:249]: + ".0" ?

File: ibm/flex/structures_test.go (hunk 1 above)
Comment (michaelkad) [reply ibm/flex/structures_test.go]: Debugging!

File: ibm/service/power/resource_ibm_pi_host.go (hunk 2 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_host.go:248]: What is the question?

Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_host.go:248]: I don't get why we are adding the .0 to the displayname when we get it.

File: website/docs/r/pi_host.html.markdown [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     Nested schema for `capacity`:
+  * `available_core` - (Float) Number of cores currently available.
+  * `available_memory` - (Float) Amount of memory currently available (in GB).
+  * `reserved_core` - (Float) Number of cores reserved for system use.
+  * `reserved_memory` - (Float) Amount of memory reserved for system use (in GB).
+  * `total_core` - (Float) Total number of cores of the host.
+  * `total_memory` - (Float) Total amount of memory of the host (in GB).
+  * `used_core` - (Float) Number of cores in use on the host.
+  * `used_memory` - (Float) Amount of memory used on the host (in GB).
+* `host_id` - (String) ID of the host.
```
Comment (ismirlia) [inline website/docs/r/pi_host.html.markdown:81]: host_id should be after host_group

File: ibm/service/power/resource_ibm_pi_host.go (hunk 2 above)
Comment (michaelkad) [reply ibm/service/power/resource_ibm_pi_host.go:242]: Getting the first element equivalent.

File: website/docs/r/pi_host.html.markdown [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    pi_cloud_instance_id = "<value of the cloud_instance_id>"
+    pi_host {
+            display_name = "<value of the display_name>"
+            sys_type = "<value of the sys_type>"
+    }
+    pi_host_group_id = "<value of the host_group_id>"
+}
+```
+
+## Notes
```
Comment (ismirlia) [inline website/docs/r/pi_host.html.markdown:28]:
```suggestion
### Notes
```

File: ibm/service/power/resource_ibm_pi_host.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+							ForceNew:    true,
+							Required:    true,
+							Description: "System type.",
+						},
+					},
+				},
+				Required: true,
+				Type:     schema.TypeList,
+			},
+			//Attributes
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:68]:
```suggestion
			// Attributes
```

File: ibm/service/power/resource_ibm_pi_host.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+		return diag.FromErr(err)
+	}
+	d.SetId("")
+
+	return nil
+}
+func isWaitForPIHostDeleted(ctx context.Context, client *instance.IBMPIHostGroupsClient, id string, timeout time.Duration) (interface{}, error) {
+	log.Printf("Waiting for host (%s) to be deleted.", id)
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{helpers.PIInstanceDeleting},
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:282]:
```suggestion
		Pending:    []string{State_DELETING},
```
Replace with State variables?

File: ibm/service/power/resource_ibm_pi_host.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+		Target:     []string{helpers.PIInstanceNotFound},
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:283]:
```suggestion
		Target:     []string{State_NotFound},
```

File: ibm/service/power/resource_ibm_pi_host.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+	return stateConf.WaitForStateContext(ctx)
+}
+
+func isPIHostDeleteRefreshFunc(client *instance.IBMPIHostGroupsClient, id string) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		host, err := client.GetHost(id)
+		if err != nil {
+			if strings.Contains(err.Error(), NotFound) {
+
+				return host, helpers.PIInstanceNotFound, nil
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:299]:
```suggestion
				return host, State_NotFound, nil
```

File: ibm/service/power/resource_ibm_pi_host.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+			}
+		}
+		return host, helpers.PIInstanceDeleting, nil
```
Comment (ismirlia) [inline ibm/service/power/resource_ibm_pi_host.go:302]:
```suggestion
		return host, State_DELETING, nil
```

File: ibm/service/power/ibm_pi_constants.go [hunk 10]
```diff
@@ -284,21 +294,25 @@ const (
 	Attr_SystemPoolName                              = "system_pool_name"
 	Attr_SystemPools                                 = "system_pools"
 	Attr_Systems                                     = "systems"
-	Attr_SysType                                     = "systype"
+	Attr_SysType                                     = "sys_type"
```
Comment (ismirlia) [inline ibm/service/power/ibm_pi_constants.go:297]: I think the value of this constant can't be changed. It is being used as "systype" in the instance data source.
https://github.com/IBM-Cloud/terraform-provider-ibm/pull/5054/files

**Comment by michaelkad** [comment]
Open in main repo.

**Pull request closed without merging**